enhancements/network/user-defined-network-segmentation.md

```diff
@@ -0,0 +1,947 @@
+network. In the future, access to more services on the default network may be granted. However, that would require more
+groundwork around enforcing network policy (which is evaluated typically after service DNAT) as potentially nftables
+rules. Such work is considered a future enhancement and beyond the scope of this initial implementation.
+
+With this proposal, OVN-Kubernetes will check which network is being used for this namespace, and then only enable the
+service there. The cluster IP of the service will only be available in the network of that service, except for KAPI and
+DNS as previously explained. Host networked pods in a namespace with a user-defined primary network will also be limited
+to only accessing the cluster IP of the services for that network. Load balancer IP and nodeport services are also
+supported on user-defined networks. Service selectors are only able to select endpoints from the same namespace where the
+service exists. Services that exist before the user-defined network is assigned to a namespace will result in
+OVN-Kubernetes executing a re-sync on all services in that namespace, and updating all load balancers. Keep in mind that
+pods must not exist in the namespace when the namespace is assigned to a new network or the new network assignment will
+not be accepted by OVN-Kubernetes.
+
+Services in a user-defined network will be reachable by other namespaces that share the same network.
+
+As previously mentioned, Kubernetes API and DNS services will be accessible by all pods.
+
+Endpoint slices will provide the IPs of the cluster default network in Kubernetes API. For this implementation the required
+endpoints are those IP addresses which reside on the user-defined primary network. In order to solve this problem,
+OVN-Kubernetes may create its own endpoint slices or may choose to do dynamic lookups at runtime to map endpoints to
+their primary IP address. Leveraging a second set of endpoint slices will be the preferred method, as it creates less
+indirection and gives explicit Kube API access to what IP addresses are being used by OVN-Kubernetes.
+
+Kubelet health checks to pods are queried via the cluster default network. When endpoints are considered unhealthy they
+will be removed from the endpoint slice, and thus their primary IP will be removed from the OVN load balancer. However,
+it is important to note that the healthcheck is being performed via the cluster default network interface on the pod, 
+which ensures the application is alive, but does not confirm network connectivity of the primary interface. Therefore,
+there could be a situation where OVN networking on the primary interface is broken, but the default interface continues
+to work and reports 200 OK to Kubelet, thus rendering the pod serving in the endpoint slice, but unable to function.
+Although this is an unlikely scenario, it is good to document.
```

Comment:
For sure, this must be documented.

As a user, I'd be extremely confused if the probed reported OK but the service I'm trying to access is unable to reply. The UX really diverges from what Kubernetes hands out.

How hard is it to get the probes working on the primary network ? Can we have this as a follow-up goal ?

Reply:
There was an idea from @tssurya about using a proxy. Yeah we should have a follow-up to try to make this more clean in the future. I can add it as a future goal.

Reply:
Actually I decided not to list it as a future goal. i think it can be an engineering goal for us, but it isn't really a feature I want to advertise right now.

Reply:
Makes sense. 

Would we require a tracker somewhere to ensure we don't forget about it ?